### ISSUE TITLE
What data types should dhall-csv support/infer?

### ISSUE THREAD

**Issue by MarcosJLR** [dhall-lang/dhall-haskell#2241]
Right now the `dhall-csv` package supports only dealing with CSVs that contain a Header and transforms them to and from Lists of records. I would want to discuss the following:

* What should we do with headerless CSVs? Should we just throw an error or try to convert it to something like a List of Lists?
* Should we allow to converting Dhall Records to CSVs with only one row? (converting something of type `{ ... }` instead of type `List { ... }`)
* Should we allow converting a CSV with just one row to a Dhall Record directly? (instead of a list with just one record)

Apart from those issues, there is a problem that arises because everything in CSV is text. What I mean by this is that Text "3.14" cannot be told apart from Double 3.14. 

So, should we implement some type inference when converting CSV to Dhall or should we parse every field as `Text` unless a schema is specified?

**Comment by sjakobi** [comment]
CC @ad-si who inspired this project (https://github.com/dhall-lang/dhall-lang/issues/1165).

**Comment by sjakobi** [comment]
> * What should we do with headerless CSVs? Should we just throw an error or try to convert it to something like a List of Lists?

I think we can produce "anonymous" records in this case, with fields labeled `_1`, `_2`, etc.


> * Should we allow to converting Dhall Records to CSVs with only one row? (converting something of type `{ ... }` instead of type `List { ... }`)
> * Should we allow converting a CSV with just one row to a Dhall Record directly? (instead of a list with just one record)

Unless there's particular demand, I think we should do without such exceptions.



> Apart from those issues, there is a problem that arises because everything in CSV is text. What I mean by this is that Text "3.14" cannot be told apart from Double 3.14.
> 
> So, should we implement some type inference when converting CSV to Dhall or should we parse every field as `Text` unless a schema is specified?

I think we should eagerly convert to numeric types where possible. We could also add CLI-options to disable this behaviour.

**Comment by ad-si** [comment]
I second @sjakobi. 
Especially the point about exceptions. Importing a CSV should always yield the same data structure.

> I think we can produce "anonymous" records in this case, with fields labeled _1, _2, etc.

Maybe even include an inferred datatype in the name `col_1_natural` (or more human friendly? `Column 1 (Natural)`)

**Comment by Gabriel439** [comment]
Yeah, I think we should treat each row as a record with fields named `_1`, `_2`, …, but still perform type inference

I don't think we need to provide special treatment for 1-row CSVs, yet, but if users ask for this feature then we can revisit it later.  It's easier to add features later than to remove them

I do think that at some point we should do type-inference on row contents, but in the short term I think it's fine if the user has to provide an explicit type for `csv-to-dhall`
